File: app/src/main/java/com/inspiredandroid/linuxcommandbibliotheca/fragments/CommandManFragment.java
package com.inspiredandroid.linuxcommandbibliotheca.fragments;

import android.app.SearchManager;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.SearchView;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ExpandableListView;
import android.widget.ImageButton;

import com.google.android.gms.appindexing.Action;
import com.inspiredandroid.linuxcommandbibliotheca.CommandManActivity;
import com.inspiredandroid.linuxcommandbibliotheca.R;
import com.inspiredandroid.linuxcommandbibliotheca.adapter.ManExpandableListAdapter;
import com.inspiredandroid.linuxcommandbibliotheca.asnytasks.SearchManAsyncTask;
import com.inspiredandroid.linuxcommandbibliotheca.interfaces.OnConvertFromHtmlToSpannableListener;
import com.inspiredandroid.linuxcommandbibliotheca.misc.AppManager;
import com.inspiredandroid.linuxcommandbibliotheca.misc.FragmentCoordinator;
import com.inspiredandroid.linuxcommandbibliotheca.models.CommandPage;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.Realm;
import io.realm.RealmResults;

/**
 * Created by Simon Schubert
 */
public class CommandManFragment extends AppIndexFragment implements OnConvertFromHtmlToSpannableListener, View.OnClickListener {

    @BindView(R.id.fragment_command_man_btn_up)
    ImageButton btnUp;
    @BindView(R.id.fragment_command_man_btn_down)
    ImageButton btnDown;
    @BindView(R.id.fraggment_commandman_elv)
    ExpandableListView mList;
    private ManExpandableListAdapter mAdapter;
    private Realm mRealm;
    private String mName;
    private long mId;
    private int mCategory;
    private int mIndexesPosition;
    private String query;
    private ArrayList<Integer> indexes = new ArrayList<>();

    /**
     * Split String every partitionSize character
     *
     * @param string
     * @param partitionSize
     * @return
     */
    private static ArrayList<String> getParts(String string, int partitionSize) {
        ArrayList<String> parts = new ArrayList<>();
        int len = string.length();
        for (int i = 0; i < len; i += partitionSize) {
            parts.add(string.substring(i, Math.min(len, i + partitionSize)));
        }
        return parts;
    }

    @SuppressWarnings("deprecation")
    public static CharSequence fromHtml(String html) {
        CharSequence result;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            result = Html.fromHtml(html, Html.FROM_HTML_MODE_LEGACY);
        } else {
            result = Html.fromHtml(html);
        }
        return result;
    }

    @Override
    public String getAppIndexingTitle() {
        return mName + "(" + mCategory + ") man page";
    }

    @Override
    public Action getAppIndexingAction() {
        final Uri APP_URI = Uri.parse("android-app://com.inspiredandroid.linuxcommandbibliotheca/http/linux.schubert-simon.de/mans");
        final Uri WEB_URL = Uri.parse("http://linux.schubert-simon.de/mans/");
        return Action.newAction(Action.TYPE_VIEW, getAppIndexingTitle(), WEB_URL, APP_URI);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (!FragmentCoordinator.isTabletLayout(getActivity())) {
            setHasOptionsMenu(true);
        }

        // Get unique command mId
        Bundle b = getArguments();
        mId = b.getLong(CommandManActivity.EXTRA_COMMAND_ID);
        mName = b.getString(CommandManActivity.EXTRA_COMMAND_NAME);
        mCategory = b.getInt(CommandManActivity.EXTRA_COMMAND_CATEGORY);

        mRealm = Realm.getDefaultInstance();

        mAdapter = createAdapter();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_command_man, container, false);

        ButterKnife.bind(this, view);

        mList.setAdapter(mAdapter);

        btnUp.setOnClickListener(this);
        btnDown.setOnClickListener(this);

        return view;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        mRealm.close();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.man, menu);

        // Associate searchable configuration with the SearchView
        SearchManager searchManager = (SearchManager) getActivity().getSystemService(Context.SEARCH_SERVICE);
        MenuItem item = menu.findItem(R.id.search);
        if (Build.VERSION.SDK_INT >= 11) {
            SearchView searchView = (SearchView) item.getActionView();
            searchView.setSearchableInfo(searchManager.getSearchableInfo(getActivity().getComponentName()));

            // Associate searchable configuration with the SearchView
            searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {

                @Override
                public boolean onQueryTextSubmit(String s) {
                    return false;
                }

                @Override
                public boolean onQueryTextChange(String query) {
                    if (query.length() > 0) {
                        search(query);
                    } else {
                        resetSearchResults();
                    }
                    return true;
                }
            });
            MenuItemCompat.setOnActionExpandListener(item, new MenuItemCompat.OnActionExpandListener() {
                @Override
                public boolean onMenuItemActionExpand(MenuItem item) {
                    return true;
                }

                @Override
                public boolean onMenuItemActionCollapse(MenuItem item) {
                    resetSearchResults();
                    return true;
                }
            });
        }

        MenuItem bookmarkItem = menu.findItem(R.id.bookmark);
        bookmarkItem.setIcon(AppManager.hasBookmark(getContext(), mId) ? R.drawable.ic_bookmark_white_24dp : R.drawable.ic_bookmark_border_white_24dp);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.bookmark) {
            toogleBookmarkState();
            getActivity().supportInvalidateOptionsMenu();
            return true;
        }
        return false;
    }

    /**
     * Split long page text into child mList views
     *
     * @return
     */
    private ManExpandableListAdapter createAdapter() {

        RealmResults<CommandPage> pages = mRealm.where(CommandPage.class).equalTo(CommandPage.COMMANDID, mId).findAll();

        ArrayList<String> groups = new ArrayList<>();
        ArrayList<ArrayList<CharSequence>> child = new ArrayList<>();
        for (CommandPage page : pages) {

            groups.add(page.getTitle());

            CharSequence chars = fromHtml(page.getPage());

            ArrayList<CharSequence> pageSplit = new ArrayList<>();
            String[] tmp = chars.toString().split("\\r?\\n");
            for (String tmpSplit : tmp) {
                if (tmpSplit.length() < 600) {
                    if (!tmpSplit.isEmpty()) {
                        pageSplit.add(tmpSplit);
                    }
                } else {
                    pageSplit.addAll(getParts(tmpSplit, 600));
                }
            }
            child.add(pageSplit);
        }

        return new ManExpandableListAdapter(getActivity(), groups, child);
    }

    private void toogleBookmarkState() {
        if (AppManager.hasBookmark(getContext(), mId)) {
            AppManager.removeBookmark(getContext(), mId);
        } else {
            AppManager.addBookmark(getContext(), mId);
        }
    }

    /**
     *
     */
    private void resetSearchResults() {
        SearchManAsyncTask async = new SearchManAsyncTask(getContext(), "", mAdapter.mChild, this);
        addAsyncTask(async);
        async.execute();

        hideButton();
    }

    /**
     * Search the man page for q occurs and highlight them and jump to first occur
     *
     * @param q
     */
    private void search(String q) {
        SearchManAsyncTask async = new SearchManAsyncTask(getContext(), q, mAdapter.mChild, this);
        addAsyncTask(async);
        async.execute();
    }

    /**
     * Get line by index of occur and calculate position
     *
     * @param index
     */
    private void scrollToPosition(int index) {
        /*
        int line = tvDescription.getLayout().getLineForOffset(index);
        int lineHeight = tvDescription.getLayout().getHeight() / tvDescription.getLayout().getLineCount();
        int position = line * lineHeight;
        scrollView.scrollTo(0, position);
        */
    }

    /**
     *
     */
    private void hideButton() {
        btnUp.setVisibility(View.GONE);
        btnDown.setVisibility(View.GONE);
    }

    /**
     *
     */
    private void showButton() {
        btnUp.setVisibility(View.VISIBLE);
        btnDown.setVisibility(View.VISIBLE);
    }

    /**
     * go to next, if last then go to first
     */
    private void jumpToNextPosition() {
        mIndexesPosition--;
        if (mIndexesPosition < 0) {
            mIndexesPosition = indexes.size() - 1;
        }
        scrollToPosition(indexes.get(mIndexesPosition));
    }

    /**
     * go to previous, if smaller 0 then go to last
     */
    private void jumpToPreviousPosition() {
        mIndexesPosition++;
        if (mIndexesPosition >= indexes.size()) {
            mIndexesPosition = 0;
        }
        scrollToPosition(indexes.get(mIndexesPosition));
    }

    @Override
    public void onConvertHtmlToSpannable(ArrayList<ArrayList<CharSequence>> spannable) {
        mAdapter.mChild = spannable;
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.fragment_command_man_btn_up) {
            jumpToNextPosition();
        } else if (v.getId() == R.id.fragment_command_man_btn_down) {
            jumpToPreviousPosition();
        }
    }
}

File: app/src/main/java/com/inspiredandroid/linuxcommandbibliotheca/fragments/CommandManFragment.kt
package com.inspiredandroid.linuxcommandbibliotheca.fragments

import android.app.SearchManager
import android.content.Context
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.widget.SearchView
import android.text.Html
import android.text.Spannable
import android.text.SpannableString
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.view.*
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.appindexing.Action
import com.inspiredandroid.linuxcommandbibliotheca.BuildConfig
import com.inspiredandroid.linuxcommandbibliotheca.CommandManActivity
import com.inspiredandroid.linuxcommandbibliotheca.R
import com.inspiredandroid.linuxcommandbibliotheca.adapter.ManAdapter
import com.inspiredandroid.linuxcommandbibliotheca.misc.AppManager
import com.inspiredandroid.linuxcommandbibliotheca.misc.FragmentCoordinator
import com.inspiredandroid.linuxcommandbibliotheca.misc.SearchIndex
import com.inspiredandroid.linuxcommandbibliotheca.misc.highlightQueryInsideText
import com.inspiredandroid.linuxcommandbibliotheca.models.Command
import com.inspiredandroid.linuxcommandbibliotheca.models.CommandPage
import io.realm.Realm
import io.realm.kotlin.where
import kotlinx.android.synthetic.main.fragment_command_man.*
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import java.util.*

/**
 * Created by Simon Schubert
 */
class CommandManFragment : AppIndexFragment(), View.OnClickListener {

    private lateinit var adapter: ManAdapter
    private lateinit var realm: Realm
    private var firebaseAnalytics: FirebaseAnalytics? = null
    private var name: String = ""
    private var id: Long = 0
    private var currentIndexPosition: Int = 0
    var indexes = ArrayList<SearchIndex>()

    private fun fromHtml(html: String?): Spanned {
        return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            Html.fromHtml(html, Html.FROM_HTML_MODE_LEGACY)
        } else {
            Html.fromHtml(html)
        }
    }

    override fun getAppIndexingAction(): Action {
        return Action.Builder(Action.Builder.VIEW_ACTION)
                .setObject("$name Linux man page", "http://linuxcommandlibrary.com/man/$name.html")
                .build()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (!FragmentCoordinator.isTabletLayout(activity)) {
            setHasOptionsMenu(true)
        }

        id = arguments?.getLong(CommandManActivity.EXTRA_COMMAND_ID) ?: 0L

        realm = Realm.getDefaultInstance()

        adapter = createAdapter()

        context?.let {
            firebaseAnalytics = FirebaseAnalytics.getInstance(it)
        }

        val realm = Realm.getDefaultInstance()
        val command = realm.where<Command>().equalTo(Command.ID, id).findFirst()
        if (command != null) {
            trackSelectContent(command.name)
            name = command.name
            activity?.title = name
        }
        realm.close()
    }

    private fun trackSelectContent(name: String?) {
        if (BuildConfig.DEBUG) {
            return
        }
        val bundle = Bundle()
        bundle.putString(FirebaseAnalytics.Param.ITEM_ID, name)
        bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "Manual Detail")
        firebaseAnalytics?.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_command_man, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView.adapter = adapter
        fastScroller.setRecyclerView(recyclerView)

        btnUp.setOnClickListener(this)
        btnDown.setOnClickListener(this)
    }

    override fun onDestroy() {
        super.onDestroy()

        realm.close()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.search_bookmark, menu)

        val item = menu?.findItem(R.id.search)
        val searchView = item?.actionView as SearchView

        activity?.let {
            val searchManager = it.getSystemService(Context.SEARCH_SERVICE) as SearchManager
            searchView.setSearchableInfo(searchManager.getSearchableInfo(it.componentName))

            searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {

                override fun onQueryTextSubmit(s: String): Boolean {
                    return false
                }

                override fun onQueryTextChange(query: String): Boolean {
                    if (!isAdded) {
                        return true
                    }
                    if (query.isNotEmpty()) {
                        search(query)
                    } else {
                        resetSearchResults()
                    }
                    return true
                }
            })
            item.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
                override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                    return true
                }

                override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                    resetSearchResults()
                    return true
                }
            })
        }

        val bookmarkItem = menu.findItem(R.id.bookmark)
        bookmarkItem.setIcon(if (AppManager.hasBookmark(context, id)) R.drawable.ic_bookmark_black_24dp else R.drawable.ic_bookmark_border_black_24dp)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == R.id.bookmark) {
            toggleBookmark()
            activity?.invalidateOptionsMenu()
            return true
        }
        return false
    }

    private fun createAdapter(): ManAdapter {
        val pages = realm.where<CommandPage>().equalTo(CommandPage.COMMANDID, id).findAll()

        val groups = ArrayList<String>()
        val child = ArrayList<ArrayList<CharSequence>>()
        for (page in pages) {

            // Split description by <p> elements to increase the recyclerview performance
            val pageSplit = ArrayList<CharSequence>()
            val tmp = page.page.toString().split("</p>".toRegex()).toTypedArray()
            for (tmpSplit in tmp) {
                val p = trim(fromHtml("$tmpSplit</p>"))
                if (p.isNotEmpty()) {
                    pageSplit.add(p)
                }
            }

            // If the parts are still to long split them Every 3000 characters
            val descriptions = ArrayList<CharSequence>()
            val size = 2000
            pageSplit.forEach {
                var index = 0
                while (index < it.length) {
                    var endIndex = index + size
                    if (endIndex >= it.length) {
                        endIndex = it.length - 1
                    }
                    descriptions.add(it.subSequence(index, endIndex))
                    index += size
                }
            }

            if (pageSplit.size > 0) {
                groups.add(page.title)
                child.add(descriptions)
            }
        }

        return ManAdapter(groups, child)
    }

    private fun trim(data: CharSequence): CharSequence {
        var text = data
        if (text.isEmpty()) {
            return text
        }
        while (text[text.length - 1] == '\n') {
            text = text.subSequence(0, text.length - 1)
        }
        while (text[0] == '\n') {
            text = text.subSequence(1, text.length)
        }
        return text
    }

    private fun toggleBookmark() {
        if (AppManager.hasBookmark(context, id)) {
            AppManager.removeBookmark(context, id)
        } else {
            AppManager.addBookmark(context, id)
        }
    }

    /**
     *
     */
    private fun resetSearchResults() {
        search("")
    }

    /**
     * Search the search_bookmark page for q occurs and highlight them and jump to first occur
     *
     * @param q
     */
    private fun search(q: String) {
        doAsync {
            indexes.clear()
            adapter.expandAll()
            var pos = 0
            adapter.parts.forEachIndexed { index, children ->
                pos++
                for (i in children.indices) {
                    val child = children[i]
                    val searchResult = child.toString().highlightQueryInsideText(context, q)
                    children[i] = searchResult.result
                    searchResult.indexes.forEach {
                        /*
                        if(indexes.count() == currentIndexPosition){
                            val ind = it
                            context?.let {
                                val span = SpannableString(children[i])
                                span.setSpan(ForegroundColorSpan(ContextCompat.getColor(it, R.color.accent_material_dark)), ind, ind+q.count(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                                children[i] = span
                            }
                        }
                        */
                        indexes.add(SearchIndex(pos, it))
                    }
                    pos++
                }
            }
            uiThread {
                tvSearchCount.text =  "$currentIndexPosition/${indexes.count()}"
                adapter.notifyDataSetChanged()
            }
        }
        if(q.isNotEmpty()) {
            showButton()
        } else {
            hideButton()
        }
    }

    /**
     * Get line by index of occur and calculate position
     *
     * @param index
     */
    private fun scrollToPosition(index: SearchIndex) {
        /*
        int line = tvDescription.getLayout().getLineForOffset(index);
        int lineHeight = tvDescription.getLayout().getHeight() / tvDescription.getLayout().getLineCount();
        int position = line * lineHeight;
        scrollView.scrollTo(0, position)
        */
        recyclerView.layoutManager.scrollToPosition(index.pos)
        tvSearchCount.text =  "$currentIndexPosition/${indexes.count()}"
        /*
        var items = adapter.items[currentIndexPosition]
        adapter.parts[items.groupId]
        */
    }

    /**
     *
     */
    private fun hideButton() {
        btnUp.visibility = View.GONE
        btnDown.visibility = View.GONE
        tvSearchCount.visibility = View.GONE
        imageView7.visibility = View.GONE
    }

    /**
     *
     */
    private fun showButton() {
        btnUp.visibility = View.VISIBLE
        btnDown.visibility = View.VISIBLE
        tvSearchCount.visibility = View.VISIBLE
        imageView7.visibility = View.VISIBLE
    }

    /**
     * go to next, if last then go to first
     */
    private fun jumpToNextPosition() {
        currentIndexPosition--
        if (currentIndexPosition < 0) {
            currentIndexPosition = indexes.size - 1
        }
        scrollToPosition(indexes[currentIndexPosition])
    }

    /**
     * go to previous, if smaller 0 then go to last
     */
    private fun jumpToPreviousPosition() {
        currentIndexPosition++
        if (currentIndexPosition >= indexes.size) {
            currentIndexPosition = 0
        }
        scrollToPosition(indexes[currentIndexPosition])
    }

    override fun onClick(v: View) {
        if (v.id == R.id.btnUp) {
            jumpToNextPosition()
        } else if (v.id == R.id.btnDown) {
            jumpToPreviousPosition()
        }
    }
}